doc/api/esm.md

```diff
@@ -1333,20 +1331,20 @@ _defaultEnv_ is the conditional environment name priority array,
 >       1. Return _resolved_.
 > 1. Throw a _Module Not Found_ error.
 
-**ESM_FORMAT**(_url_, _isMain_)
+**ESM_FORMAT**(_url_)
 
-> 1. Assert: _url_ corresponds to an existing file.
+> 1. Assert: _url_ corresponds to an existing file pathname.
 > 1. Let _pjson_ be the result of **READ_PACKAGE_SCOPE**(_url_).
 > 1. If _url_ ends in _".mjs"_, then
 >    1. Return _"module"_.
 > 1. If _url_ ends in _".cjs"_, then
 >    1. Return _"commonjs"_.
 > 1. If _pjson?.type_ exists and is _"module"_, then
->    1. If _isMain_ is **true** or _url_ ends in _".js"_, then
+>    1. If _url_ ends in _".js"_ or lacks a file extension, then
```

Comment:
We specifically don't want to support loading files without an extension that are not the main though.

Reply:
Can you specify why / since you can import the main. Right now we do support loading unknown file extensions as main so that seems to go against 

> One of the major features of the ES module resolver is being able to support new file extensions in future, which is reliant on the fact that we always throw for unknown file extensions.

Reply:
It is an exception in the resolver specifically to support `./bin/pkg` in the ES module resolver where the bin file has no extension and uses a hash bang.

Otherwise in a `"type": "module"` package these bin use cases are no longer supported.

But supporting the above workflows should not allow us to block forwards compatibility for new file extensions in future.

Reply:
I'm not sure the original code is acting as intended if the idea is to provide forwards compatibility.

See within a package of `"type":"module"` how `node --experimental-wasm-modules ./foo.wasm` loads as WASM, but without `--experimental-wasm-modules` loads as ESM. It seems for forwards compatibility it should error without `--experimental-wasm-modules` and certainly not flip from non-error behavior to different behavior when that flag is passed.

Reply:
I previously missed that this explicitly forbids unknown extensions so does not block future extensions. Marking as resolved.